chain-core/src/common/fixed.rs

```diff
@@ -25,45 +26,196 @@ fn continued_fraction(
 }
 
 // https://en.wikipedia.org/wiki/Exponential_function#Continued_fractions_for_ex
-fn exp_series(
+fn exp2_series(
     x: FixedNumber,
+    y: FixedNumber,
 ) -> (
     impl Iterator<Item = FixedNumber>,
     impl Iterator<Item = FixedNumber>,
 ) {
-    let it_a = vec![FixedNumber::from_num(1), FixedNumber::from_num(1)]
+    let x2 = x * x;
+    let it_a = vec![FixedNumber::from_num(1), y * 2 - x]
         .into_iter()
-        .chain((2..).map(move |n| x + FixedNumber::from_num(n)));
-    let it_b = iter::once(1).chain((1..).map(|n| -n)).map(move |n| x * n);
+        .chain((0..).map(move |i| y * (i * 4 + 6)));
+    let it_b = iter::once(2 * x).chain(iter::repeat(x2));
     (it_a, it_b)
 }
 
 /// Compute exponencial function with continued fraction method
-pub fn exp(x: FixedNumber) -> FixedNumber {
-    continued_fraction(exp_series(x), 100)
+pub fn exp2(x: FixedNumber, y: FixedNumber) -> FixedNumber {
+    continued_fraction(exp2_series(x, y), 25)
+}
+
+// https://en.wikipedia.org/wiki/Natural_logarithm#Continued_fractions
+fn log2_series(
+    x: FixedNumber,
+    y: FixedNumber,
+) -> (
+    impl Iterator<Item = FixedNumber>,
+    impl Iterator<Item = FixedNumber>,
+) {
+    let n1 = 2 * y + x;
+    let x2 = x * x;
+    let it_a = iter::once(FixedNumber::from_num(0)).chain((1..).step_by(2).map(move |i| n1 * i));
+    let it_b = iter::once(2 * x).chain((1..).map(move |i| -x2 * (i * i)));
+    (it_a, it_b)
+}
+
+/// Compute natural logarithm function with continued fraction method
+pub fn log2(x: FixedNumber, y: FixedNumber) -> FixedNumber {
+    continued_fraction(log2_series(x, y), 10)
+}
+
+pub fn monetary_expansion(staking: u64, tau: u64, r0: u64, period: u64) -> u64 {
```

Comment:
perhaps some top-level comment / explanation + richer types than `u64` (either type aliases or opaque types) could make it a bit clearer and less-error prone (so that one doesn't pass in wrong arguments in wrong places; or if it's refactored) 

e.g. `staking` / `staking_` could be `Coin`? I guess?
period could be some `Timespec` or some clarification it's in seconds?
`tau` and `r0` could be directly `FixedNumber` or some opaque type that captures their safe ranges -- e.g. have `pub struct Tau` that guarantees range ` [1, 100_00000000_00000000]`, `pub struct R0` ([0, 1000]) ...

Reply:
those opaque types could be used in network parameters as well to make it consistent

Reply:
The return type could perhaps be `Coin`?

Reply:
I've changed the staking parameter and return type to `Coin`, `r0` to `Millis`. Inside the computation they are all converted to `Fixed`.
I'm not sure about using `Timespec` for `period`, because I remember not long ago, `Timespec` is still a signed type, maybe better to use `u64` directly here.

Reply:
And `tau`/`r0`/`period ` has no inherent upper range for this formula, except some too big value might make the computation overflow.